Implement void[]-based ring queue in Mirror handler

gavin-norman-sociomantic opened the pull request:
Fixes two issues:
1. We encountered segfaults under high load. Investigation revealed that
   this is due to casting a `void[]*` to a pointer to another type of array
   (as we were doing in the old queue implementation). The exact cause of
   the segfaults was not discovered, but it appears to be related to GC
   corruption.
   See https://github.com/sociomantic-tsunami/dhtnode/issues/15.
2. The previous queue implementation was pretty inefficient: it required a
   memory shift of the entire queue after each pop. The new implementation
   does not require any memory shifting.

gavin-norman-sociomantic commented on src/dhtproto/node/neo/request/Mirror.d:
```diff
@@ -260,7 +371,8 @@ public abstract scope class MirrorProtocol_v0
 
     ***************************************************************************/
-
+import ocean.io.Stdout;
```
Haha, ok I'll remove this debug stuff ;)

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Mirror.d:
```diff
@@ -65,18 +69,50 @@ public abstract scope class MirrorProtocol_v0
+        {
+            this.max_elems = (max_size / E.sizeof) * E.sizeof;
+            buf.length = this.max_elems * E.sizeof;
+            this.queue = (cast(E*)buf.ptr)[0 .. this.max_elems];
```
Suggest
- `ref void[] buf` now that you don't store a pointer to the array any more,
- `(cast(E[])*buf)[0 .. this.max_elems]` (without the `*` for `ref void[] buf`).

david-eckardt-sociomantic replied:
Actually `this.queue = cast(E[])buf;` should be sufficient

gavin-norman-sociomantic replied:
Updated.

Commit:
Implement void[]-based ring queue in Mirror handler

Fixes two issues:
1. We encountered segfaults under high load. Investigation revealed that
   this is due to casting a `void[]*` to a pointer to another type of array
   (as we were doing in the old queue implementation). The exact cause of
   the segfaults was not discovered, but it appears to be related to GC
   corruption.
   See https://github.com/sociomantic-tsunami/dhtnode/issues/15.
2. The previous queue implementation was pretty inefficient: it required a
   memory shift of the entire queue after each pop. The new implementation
   does not require any memory shifting.

joseph-wakeling-sociomantic commented:
Can confirm that this (in combination with swarm changes in v4.2.4) results in no segfaults during a mirror request for my use-case.
